ledger/store/trackerdb/pebbledbdriver/pebbledriver.go

```diff
@@ -0,0 +1,572 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package pebbledbdriver
+
+import (
+	"context"
+	"io"
+	"runtime"
+	"time"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/ledger/store/trackerdb"
+	"github.com/algorand/go-algorand/ledger/store/trackerdb/generickv"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/util/db"
+	"github.com/cockroachdb/pebble"
+	"github.com/cockroachdb/pebble/bloom"
+	"github.com/cockroachdb/pebble/vfs"
+)
+
+const (
+	// minCache is the minimum amount of memory in megabytes to allocate to pebble
+	// read and write caching, split half and half.
+	minCache = 16
+
+	// minHandles is the minimum number of files handles to allocate to the open
+	// database files.
+	minHandles = 16
+)
+
+type trackerStore struct {
+	kvs   kvstore
+	proto config.ConsensusParams
+	// use the generickv implementations
+	trackerdb.Reader
+	trackerdb.Writer
+	trackerdb.Catchpoint
+}
+
+// Open opens a Pebble db database
+func Open(dbdir string, inMem bool, proto config.ConsensusParams, log logging.Logger) (trackerdb.Store, error) {
+	cache := 128 // this divided by 2 and by memTableLimit = 1GB /(2 * 16) = 32MB per memtable
+	handles := 1000
+
+	// Ensure we have some minimal caching and file guarantees
+	if cache < minCache {
+		cache = minCache
+	}
+	if handles < minHandles {
+		handles = minHandles
+	}
+
+	// The max memtable size is limited by the uint32 offsets stored in
+	// internal/arenaskl.node, DeferredBatchOp, and flushableBatchEntry.
+	// Taken from https://github.com/cockroachdb/pebble/blob/master/open.go#L38
+	maxMemTableSize := 4<<30 - 1 // Capped by 4 GB
+
+	memTableLimit := 2 // default: 2
+	memTableSize := cache * 1024 * 1024 / 2 / memTableLimit
+	if memTableSize > maxMemTableSize {
+		memTableSize = maxMemTableSize
+	}
+
+	// configure pebbledb
+	opts := &pebble.Options{
```

Comment:
Seeing as this would be the first published pebble options published to mainline, it may be setting some precedents. In theory a user could flip to use the KV store, and then get locked into this options set, and find themselves incompatible if/when we develop these options further.

I think the situation in specific is not much of an issue. That user is consuming an experimental branch, and getting experimental results. However, we should still take a moment (like, here in this PR) to evaluate what sort of 1-way-doors we're going through when committing this driver.

I guess I'm thinking -- how can we keep from this becoming a configuration we're too afraid to change?